File: labs/favoritos/src/java/Solver.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;
import java.util.stream.Collectors;

class Fone {
    private String id;
    private String number;
    public Fone(String id, String number) {
        this.id = id;
        this.number = number;
    }

    //verifica se o número é um número de telefone válido
    public static boolean validate(String number) {
        String validos = "()-.0123456789";
        for(int i = 0; i < number.length(); i++)
            if(!validos.contains("" + number.charAt(i)))
                return false;
        return true;
    }

    //O resultado deve ficar assim
    //oi:1234
    @Override
    public String toString() {
        return getId()+ ":" + getNumber();
    }

    //GETS e SETS
    String getId() {
        return this.id;
    }
    void setId(String id) {
        this.id = id;
    }
    String getNumber() {
        return this.number;
    }
    void setNumber(String number) {
        this.number = number;
    }
}

class Contact {
    private String name;
    private List<Fone> fones;
    protected String prefix = "-"; //utilizado no toString
    private boolean star;
    //Crie um ArrayList para o ATRIBUTO fones
    //Se a variável fones não for null, adicione todos os fones usando o método addFone
    //Inicie star como false
    public Contact(String name, List<Fone> fones) {
        this.star = false;
        this.name = name;
        this.fones = new ArrayList<>();
        if(fones != null)
            fones.stream().forEach((fone) -> {this.addFone(fone);});
    }

    //Se fone for válido, insira no atributo fones
    //Se não, informe o erro
    public void addFone(Fone fone) {
        if(Fone.validate(fone.getNumber()))
            this.fones.add(fone);
        else
            System.out.println("fail: invalid number");
    }

    //Se o índice existir no ArrayList, remova o telefone com esse índice
    public void rmFone(int index) {
        if(index < 0 || index >= (int) fones.size())
            return;
        fones.remove(index);
    }

    //Use um contador para mostrar o índice do telefone
    //Use o toString do fone para adicioná-lo à saída
    //O resultado dever ficar assim:
    //- david [0:oi:123] [1:tim:9081] [2:claro:5431]
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append(this.prefix + " " + this.getName());
        for(int i = 0; i < fones.size(); i++)
            sb.append( " [" + i + ":" + fones.get(i) + "]");
        return sb.toString();
    }

    //GETS e SETS
    String getName() {
        return this.name;
    }
    void setName(String name) {
        this.name = name;
    }
    List<Fone> getFones() {
        return this.fones;
    }
    //muda o prefixo e o valor de star
    void setStar(boolean value) {
        this.star = value;
        prefix = value ? "@" : "-";
    }
    boolean getStar() { 
        return this.star;
    }
}

class Agenda {
    private List<Contact> contacts;

    public Agenda() {
        this.contacts = new ArrayList<>();
    }

    //retorna a posição do contato com esse nome no vetor ou -1 se não existir.
    private int findPos(String name) {
        for(int i = 0; i < this.contacts.size(); i++)
            if(contacts.get(i).getName().equals(name))
                return i;
        return -1;
    }

    //retorna o objeto contato com esse nome ou null se não existir
    //utilize o método findPos
    public Contact getContact(String name) {
        int pos = this.findPos(name);
        if(pos == -1)
            return null;
        return contacts.get(pos);
    }

    //se nenhum contato existir com esse nome, adicione
    //se ja existir, faça o merge adicionando os telefones
    public void addContact(Contact contact) {
        if(contact == null)
            return;
        Contact old = getContact(contact.getName());
        if(old == null) {
            this.contacts.add(contact);
        } else {
            contact.getFones().stream().forEach(fone -> old.addFone(fone));
        }
        this.contacts.sort((Contact c1, Contact c2) -> (c1.getName().compareTo(c2.getName())));
        // this.contacts.sort(new Comparator<Contact>() {
        //     @Override
        //     public int compare(Contact arg0, Contact arg1) {
        //         return ((Contact) arg0).getName().compareTo(((Contact)arg1).getName());
        //     }
        // });
    }

    //Utilize o método findPos
    public void rmContact(String name) {
        int pos = findPos(name);
        if(pos == -1) {
            System.out.println("fail: contact not found");
            return;
        }
        this.contacts.remove(pos);
    }

    //Monte uma lista auxiliar procurando no .toString() de cada contato
    //se ele possui a substring procurada.
    public List<Contact> search(String pattern) {
        return this.contacts
            .stream()
            .filter(contact -> contact.toString().contains(pattern))
            .collect(Collectors.toList());
    }



    @Override
    public String toString() {
        return this.contacts.stream().map(contact -> "" + contact).collect(Collectors.joining("\n"));
    }

    //se o contato existir, altere o atributo star dele
    public void star(String name, boolean value) { 
        Contact contact = getContact(name);
        if(contact == null)
            return;
        contact.setStar(value);
    }

    //filtre em uma nova lista apenas os contatos que estão favoritados
    public List<Contact> getStarred() {
        return this.contacts.stream().filter(c -> c.getStar()).collect(Collectors.toList());
    }

    List<Contact> getContacts() {
        return this.contacts;
    }
}

//!KEEP
class Solver {
    //cria um contato a partir do vetor de entrada tal como
    //add joao oi:123 tim:432 claro:09123
    static Contact parseContact(String[] ui) {
        return new Contact(ui[1], Arrays.asList(ui).stream()
            .skip(2).map(token -> new Fone(token.split(":")[0], token.split(":")[1]))
            .collect(Collectors.toList()));
    }
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Agenda agenda = new Agenda();
        while(true) {
            String line = scanner.nextLine();
            System.out.println("$" + line);
            String ui[] = line.split(" ");
            
            if(ui[0].equals("end")) {
                break;
            } else if(ui[0].equals("init")) {
                agenda = new Agenda();
            } else if(ui[0].equals("add")) { //name label:fone label:fone label:fone
                agenda.addContact(Solver.parseContact(ui));
            } else if(ui[0].equals("rm")) { //name
                agenda.rmContact(ui[1]);;
            } else if(ui[0].equals("rmFone")) { //name index
                agenda.getContact(ui[1]).rmFone(Integer.parseInt(ui[2]));
            } else if(ui[0].equals("show")) {
                System.out.println(agenda);
            } else if(ui[0].equals("star")) {
                agenda.star(ui[1], true);
            } else if(ui[0].equals("unstar")) {
                agenda.star(ui[1], false);
            } else if(ui[0].equals("starred")) {
                System.out.println(agenda.getStarred().stream().map(c -> "" + c).collect(Collectors.joining("\n")));
            } else if(ui[0].equals("search")) {
                System.out.println(agenda.search(ui[1]).stream().map(c -> "" + c).collect(Collectors.joining("\n")));
            }
            else {
                System.out.println("fail: invalid command");
            }
        }
        scanner.close();
    }
}

class Manual {
    public static void main(String[] args) {
        Agenda agenda = new Agenda();
        agenda.addContact(new Contact("eva", Arrays.asList(new Fone("oio", "8585"), new Fone("cla", "9999"))));
        agenda.addContact(new Contact("ana", Arrays.asList(new Fone("Tim", "3434"))));
        agenda.addContact(new Contact("bia", Arrays.asList(new Fone("viv", "5454"))));
        agenda.addContact(new Contact("ana", Arrays.asList(new Fone("cas", "4567"), new Fone("oio", "8754"))));
        System.out.println(agenda);
        /*
        - ana [0:casa:4567][1:oi:8754]
        - bia [0:vivo:5454]
        - eva [0:oi:8585][1:claro:9999]
        - rui [0:casa:3233]
        - zac [0:fixo:3131]
        */

        //case favoritando
        agenda.star("eva", true);
        agenda.star("ana", true);
        agenda.star("ana", true);
        agenda.star("zac", true);
        agenda.star("rex", true);
        //fail: contato rex nao existe
        System.out.println(agenda);
        /*
        @ ana [0:casa:4567][1:oi:8754]
        - bia [0:vivo:5454]
        @ eva [0:oi:8585][1:claro:9999]
        - rui [0:casa:3233]
        @ zac [0:fixo:3131]
        */

        //case lista de favoritos
        for(Contact fav : agenda.getStarred())
            System.out.println(fav);
        /*
        @ ana [0:casa:4567][1:oi:8754]
        @ eva [0:oi:8585][1:claro:9999]
        @ zac [0:fixo:3131]
        */

        //case removendo contato
        agenda.rmContact("zac");
        System.out.println(agenda);
        /*
        @ ana [0:casa:4567][1:oi:8754]
        - bia [0:vivo:5454]
        @ eva [0:oi:8585][1:claro:9999]
        - rui [0:casa:3233]
        */
        for(Contact fav : agenda.getStarred())
            System.out.println(fav);
        /*
        @ ana [0:casa:4567][1:oi:8754]
        @ eva [0:oi:8585][1:claro:9999]
        */

        //case desfavoritando
        agenda.star("ana", false);
        for(Contact fav : agenda.getStarred())
            System.out.println(fav);
        //@ eva [0:oi:8585][1:claro:9999]
        System.out.println(agenda);
        /*
        - ana [0:casa:4567][1:oi:8754]
        - bia [0:vivo:5454]
        @ eva [0:oi:8585][1:claro:9999]
        - rui [0:casa:3233]
        */
    }
}
//!OFF
    

File: labs/ligacao/src/cpp/solver.cpp
#include <iostream>
#include <vector>
#include <sstream>
#include <map>

using namespace std;

class Fone{
public:
    string id;
    string number;

    Fone(string serial = ""){
        stringstream ss(serial);
        getline(ss, this->id, ':');
        ss >> this->number;
    }
    
    Fone(string id, string number){
        this->id = id;
        this->number = number;
    }

    static bool validar(string number){
        string valid = "0123456789()-.";
        for(char c : number)
            if(valid.find(c) == string::npos)
                return false;
        return true;
    }
    friend ostream& operator<<(ostream& out, const Fone& fone){
        out << fone.id << ":" << fone.number;
        return out;
    }
};

class Contact {
public:
    string id;
    vector<Fone> fones;
    bool starred;
    int calls;

    Contact(string id = string(""), vector<Fone> fones= {}) : 
        id{id}, fones{fones}, starred{false}, calls{0} {
    }

    void addFone(Fone fone){
        fones.push_back(fone);
    }

    void rmFone(int index){
        fones.erase(fones.begin() + index);
    }

    vector<Fone> getAllFones(){
        return fones;
    }

    bool matchNumber(string number){
        for(Fone fone : fones)
            if(fone.number == number)
                return true;
        return false;
    }

    void increaseCallCount(){
        this->calls += 1;
    }

    friend ostream& operator<<(ostream& out, Contact contact){
        out << (contact.starred ? "@ " : "- ");
        out << contact.id;
        if(contact.calls > 0)
            out << " {" << contact.calls << " call}";
        int i = 0;
        for(auto fone: contact.fones)
            out << " [" << i++ << ":" << fone << "]";
        return out;
    }
};

template <class T>
string toString(T t){
    stringstream ss;
    ss << t;
    return ss.str();
}

#include <algorithm>

class Agenda {
    map<string, Contact*> contacts; //vetor
    map<string, Contact*> bookmarks;

public:
    ~Agenda(){
        for(auto& par : contacts)
            delete par.second;
    }

    void addContact(string id, vector<Fone> fones){
        if(contacts.count(id) == 1){
            for(auto fone : fones)
                contacts[id]->addFone(fone);
            return;
        }
        contacts[id] = new Contact(id, fones);
    }

    void rmContact(string nome) {
        auto contact = getContact(nome);
        if(contact != nullptr){
            contacts.erase(nome);
            bookmarks.erase(nome);
            delete contact;
        }
    }

    Contact * getContact(string nome){
        auto it = contacts.find(nome);
        if(it == contacts.end())
            return nullptr;
        return it->second;
    }

    void star(string nome) {
        Contact * contact = getContact(nome);
        if(contact == nullptr){
            throw "fail: contato " + nome + " nao existe";
            return;
        }
        if(contact->starred)
            return;
        contact->starred = true;
        bookmarks[nome] = contact;
    }

    void unstar(string nome){
        if(bookmarks.count(nome) == 0)
            return;
        bookmarks[nome]->starred = false;
        bookmarks.erase(nome);
    }

    vector<Contact> getStarred(){
        vector<Contact> resp;
        for(auto par : bookmarks)
            resp.push_back(*par.second);
        return resp;
    }

    vector<Contact*> getContacts(){
        vector<Contact*> resp;
        for(auto par : contacts)
            resp.push_back(par.second);
        return resp;
    }

    vector<Contact> search(string pattern){
        vector<Contact> resp;
        for(auto par : contacts){
            if(toString(*par.second).find(pattern) != string::npos)
                resp.push_back(*par.second);
        }
        return resp;
    }
};


class CallManager{
    Agenda & agenda;
    CallManager(Agenda & ag): agenda{ag}{
    }

    vector<Contact*> searchByNumber(string number){
        vector<Contact*> output;
        for(Contact * contact : agenda.getContacts())
            if(contact->matchNumber(number))
                output.push_back(contact);
        return output;
    }

    void call(string number){
        auto matches = searchByNumber(number);
        if(matches.size() > 0)
            matches[0]->increaseCallCount();
        
    }
}

Fone format_fone(string sfone){
    stringstream ss(sfone);
    string id, fone;
    getline(ss, id, ':');
    ss >> fone;
    return Fone(id, fone);
}

template <class T>
T get(stringstream& ss){
    T value;
    ss >> value;
    return value;
}

int main(){
    Agenda agenda;
    while(true){
        string line, cmd;
        getline(cin, line);
        stringstream ss(line);
        ss >> cmd;
        cout << "$" << line << endl;
        try{
            if(cmd == "end"){
                break;
            }else if(cmd == "add"){ //id label-fone label-fone label-fone
                string name = get<string>(ss);
                vector<Fone> fones;
                string fonestr;
                while(ss >> fonestr)
                    fones.push_back(format_fone(fonestr));
                agenda.addContact(name, fones);
            }else if(cmd == "rmContact"){
                agenda.rmContact(get<string>(ss));
            }else if(cmd == "rmFone"){
                Contact * cont = agenda.getContact(get<string>(ss));
                cont->rmFone(get<int>(ss));
            }else if(cmd == "show"){
                for(auto elem : agenda.getContacts())
                    cout << elem << "\n";
            }else if(cmd == "search"){
                for(auto elem : agenda.search(get<string>(ss)))
                    cout << elem << "\n";
            }else if(cmd == "star"){
                agenda.star(get<string>(ss));
            }else if(cmd == "unstar"){
                agenda.unstar(get<string>(ss));
            }else if(cmd == "starred"){
                for(auto elem : agenda.getStarred())
                    cout << elem << "\n";
            }else{
                cout << "fail: comando invalido\n";
            }
        }catch(string e){
            cout << e << endl;
        }
    }
}